### PR TITLE
Remote Teleporter Control

### PR DESCRIPTION
**Pull request by Geevies**
* Added a laptop in the teleporter room, that has a remote teleporter control program loaded.
* Removed teleportation computers, their functionality have been shoved into the stations.
* Fixed teleportation pads no longer need to be test fired to be accurate.

### CODE REVIEW

**Comment by Geevies** [comment]
!review

File: code/modules/modular_computers/file_system/programs/app_presets.dm [hunk 1]
```diff
@@ -234,6 +234,27 @@
 	)
 	return _prg_list
 
+/datum/modular_computer_app_presets/command/teleporter
+	name = "command_teleporter"
+	display_name = "Command - Teleporter"
+	description = "Contains the most common command programs and has a special teleporter control program loaded."
+	available = FALSE
+
```
Comment (Arrow768) [inline code/modules/modular_computers/file_system/programs/app_presets.dm:242]: If you want to place that laptop in a public area, dont use a command subtype.

File: code/modules/modular_computers/file_system/programs/command/teleporter.dm [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+		data["teleport_implants"] = teleport_implant_info
+
+	return data
+
+/datum/computer_file/program/teleporter/Topic(href, href_list)
+	if(..())
+		return TRUE
+
+	if(href_list["comp"])
+		linked_comp = locate(href_list["comp"]) in range(3, get_turf(computer.loc))
```
Comment (Arrow768) [inline code/modules/modular_computers/file_system/programs/command/teleporter.dm:108]: Dont use a hardref here

**Review by Arrow768 — CHANGES_REQUESTED** [review]

File: code/modules/modular_computers/file_system/programs/command/teleporter.dm [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	else if(href_list["beacon"])
+		linked_comp.locked = locate(href_list["beacon"]) in teleportbeacons
+		computer.output_message(SPAN_NOTICE("Locked in."), 1)
+		linked_comp.visible_message(SPAN_NOTICE("Locked in."), range = 2)
+	else if(href_list["implant"])
+		linked_comp.locked = locate(href_list["implant"]) in implants
+		computer.output_message(SPAN_NOTICE("Locked in."), 1)
+		linked_comp.visible_message(SPAN_NOTICE("Locked in."), range = 2)
+
+	SSvueui.check_uis_for_change(src)
```
Comment (Karolis2011) [inline code/modules/modular_computers/file_system/programs/command/teleporter.dm:118]: Does data meaningfully change? Does it need to update ui's for all users of this object that are opened? If not just make sure proc returns `TRUE`

File: code/modules/modular_computers/file_system/programs/command/teleporter.dm [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+			if(area_index[tmpname])
+				tmpname = "[tmpname] ([++area_index[tmpname]])"
+			else
+				area_index[tmpname] = 1
+			var/list/teleporter_info = list(
+				"beacon_name" = tmpname,
+				"ref" = "\ref[R]"
+				)
+			teleport_beacon_info[++teleport_beacon_info.len] = teleporter_info
+		data["teleport_beacons"] = teleport_beacon_info
```
Comment (Karolis2011) [inline code/modules/modular_computers/file_system/programs/command/teleporter.dm:73]: So vueui migth have issue with reactivity here. 
This if statement conditionally sends data what is no no for root level data
```suggestion
	data["nearby_comps"] = list()
	data["teleport_beacons"] = list()
	if(!linked_comp)
		for(var/obj/machinery/computer/teleporter/CT in range(3, T))
			var/list/comp_info = list(s
				"comp_name" = CT.name,
				"ref" = "\ref[CT]"
				)
			data["nearby_comps"][++data["nearby_comps"].len] = comp_info
	else
		data["comp_locked_in"] = !!linked_comp.locked
		var/list/area_index = list()

		var/list/teleport_beacon_info = list()
		for(var/obj/item/device/radio/beacon/R in teleportbeacons)
			var/turf/BT = get_turf(R)
			if(!BT)
				continue
			if(isNotStationLevel(BT.z))
				continue
			var/tmpname = BT.loc.name
			if(area_index[tmpname])
				tmpname = "[tmpname] ([++area_index[tmpname]])"
			else
				area_index[tmpname] = 1
			var/list/teleporter_info = list(
				"beacon_name" = tmpname,
				"ref" = "\ref[R]"
				)
			teleport_beacon_info[++teleport_beacon_info.len] = teleporter_info
		data["teleport_beacons"] = teleport_beacon_info
```

File: code/modules/modular_computers/file_system/programs/command/teleporter.dm [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		ui.auto_update_content = TRUE
+	ui.open()
+
+/datum/computer_file/program/teleporter/vueui_transfer(oldobj)
+	. = FALSE
+	var/uis = SSvueui.transfer_uis(oldobj, src, "mcomputer-command-teleporter", 600, 400, "Teleporter Control System")
+	for(var/tui in uis)
+		var/datum/vueui/ui = tui
+		ui.auto_update_content = TRUE
+		. = TRUE
```
Comment (Karolis2011) [inline code/modules/modular_computers/file_system/programs/command/teleporter.dm:27]: This enables auto update, but doesn't disable on transfer out.

**Review by Karolis2011 — CHANGES_REQUESTED** [review]

File: code/modules/modular_computers/file_system/programs/app_presets.dm (hunk 1 above)
Comment (Geevies) [reply code/modules/modular_computers/file_system/programs/app_presets.dm:242]: as discussed, this isn't in a public area

File: code/modules/modular_computers/file_system/programs/command/teleporter.dm (hunk 3 above)
Comment (Geevies) [reply code/modules/modular_computers/file_system/programs/command/teleporter.dm:128]: in testing, return TRUE didn't make it update properly

File: code/modules/modular_computers/file_system/programs/command/teleporter.dm (hunk 5 above)
Comment (Geevies) [reply code/modules/modular_computers/file_system/programs/command/teleporter.dm:27]: fixed

**Review by Karolis2011 — APPROVED** [review]
Vue file looks reasonable, tho I assume it also looks reasonable

**Review by Arrow768 — CHANGES_REQUESTED** [review]
I think that changing the locked variable to be a list that is only intendet to hold one value is a relatively hacky.
Imho it would be better to just use two variables for the same thing.

I also think that the interaction of the application with Application -> Teleporter Computer -> Teleporter Hub is pretty hacky.
Imho it would be better to get rid of the teleporter computer and move much of the functionality to the teleporter hub.
Then create a modular computer application that is intended to control the teleporter hub (without interfacing with the teleporter computer)
At the same time it would imho be worth to clean up all the old code in there. (Clown Planet, ...)

**Review by Arrow768 — CHANGES_REQUESTED** [review]
return the test-firing functionally (possibly rebranded to re-calibration)
Ideally it should:
- Take a short moment to re-calibrate
- Increase the failure-chance with each human teleported (without re-calibration)

**Review by Arrow768 — APPROVED** [review]

**Pull request merged**